File: studio/services/asset_service.py
"""Asset upload orchestration."""

import re
import uuid

from studio.repositories import assets_repo, projects_repo
from studio.storage import s3

_ALLOWED_TYPES = frozenset(assets_repo.allowed_asset_types())
_FILENAME_RE = re.compile(r"^[\w.\- ]{1,120}$", re.I)


class AssetValidationError(ValueError):
    """Raised when an asset request fails validation."""


def _require_project(project_id):
    if not projects_repo.get(project_id):
        raise ValueError("project not found")


def create_presign(project_id, *, filename, content_type, asset_type):
    _require_project(project_id)
    if asset_type not in _ALLOWED_TYPES:
        raise AssetValidationError("invalid asset type")
    if not filename or not _FILENAME_RE.match(filename):
        raise AssetValidationError("invalid filename")
    if not content_type or "/" not in content_type:
        raise AssetValidationError("invalid content type")

    asset_id = uuid.uuid4().hex[:12]
    s3_key = s3.build_asset_key(project_id, asset_id, filename)
    signed = s3.create_presigned_put(s3_key, content_type)
    assets_repo.create_pending(
        project_id,
        asset_id=asset_id,
        asset_type=asset_type,
        s3_key=s3_key,
        filename=filename,
        content_type=content_type,
    )
    return {
        "assetId": asset_id,
        "s3Key": s3_key,
        "uploadUrl": signed["url"],
        "method": signed["method"],
        "headers": signed["headers"],
        "expiresAt": signed["expiresAt"],
    }


def _with_public_url(asset):
    if not asset:
        return asset
    out = dict(asset)
    public = s3.public_asset_url(asset.get("s3Key", ""))
    if public:
        out["publicUrl"] = public
    return out


def confirm_upload(
    project_id,
    *,
    asset_id,
    s3_key,
    content_type,
    size_bytes,
    width=None,
    height=None,
):
    _require_project(project_id)
    pending = assets_repo.get(project_id, asset_id)
    if not pending:
        raise AssetValidationError("asset not found")
    if pending["s3Key"] != s3_key:
        raise AssetValidationError("s3 key mismatch")
    if pending["contentType"] != content_type:
        raise AssetValidationError("content type mismatch")
    if size_bytes <= 0:
        raise AssetValidationError("invalid file size")
    asset = assets_repo.confirm(
        project_id,
        asset_id,
        size_bytes=size_bytes,
        width=width,
        height=height,
    )
    return _with_public_url(asset)


def list_assets(project_id):
    _require_project(project_id)
    return [_with_public_url(asset) for asset in assets_repo.list_for_project(project_id)]

File: src/server/services/assetService.ts
import * as assetsRepo from "../repos/assetsRepo";
import * as projectsRepo from "../repos/projectsRepo";
import * as s3 from "../storage/s3";

const FILENAME_RE = /^[\w.\- ]{1,120}$/i;

export class AssetValidationError extends Error {}

async function requireProject(projectId: string): Promise<void> {
  if (!(await projectsRepo.get(projectId))) throw new Error("project not found");
}

export async function createPresign(
  projectId: string,
  data: { filename: string; contentType: string; assetType: string; slotKey?: string | null },
): Promise<Record<string, unknown>> {
  await requireProject(projectId);
  const allowed = new Set(assetsRepo.allowedAssetTypes());
  if (!allowed.has(data.assetType)) throw new AssetValidationError("invalid asset type");
  if (!data.filename || !FILENAME_RE.test(data.filename)) {
    throw new AssetValidationError("invalid filename");
  }
  if (!data.contentType || !data.contentType.includes("/")) {
    throw new AssetValidationError("invalid content type");
  }
  const assetId = randomId();
  const s3Key = s3.buildAssetKey(projectId, assetId, data.filename);
  const signed = await s3.createPresignedPut(s3Key, data.contentType);
  await assetsRepo.createPending(projectId, {
    assetId,
    assetType: data.assetType,
    slotKey: data.slotKey,
    s3Key,
    filename: data.filename,
    contentType: data.contentType,
  });
  return {
    assetId,
    slotKey: data.slotKey ?? null,
    s3Key,
    uploadUrl: signed.url,
    method: signed.method,
    headers: signed.headers,
    expiresAt: signed.expiresAt,
  };
}

function withPublicUrl(asset: Record<string, unknown> | null): Record<string, unknown> | null {
  if (!asset) return asset;
  const out = { ...asset };
  const pub = s3.publicAssetUrl(String(asset.s3Key ?? ""));
  if (pub) out.publicUrl = pub;
  return out;
}

export async function confirmUpload(
  projectId: string,
  data: {
    assetId: string;
    s3Key: string;
    contentType: string;
    sizeBytes: number;
    width?: number | null;
    height?: number | null;
    slotKey?: string;
    assetType?: string;
  },
): Promise<Record<string, unknown>> {
  await requireProject(projectId);
  const pending = await assetsRepo.get(projectId, data.assetId);
  if (!pending) throw new AssetValidationError("asset not found");
  if (pending.s3Key !== data.s3Key) throw new AssetValidationError("s3 key mismatch");
  if (pending.contentType !== data.contentType) throw new AssetValidationError("content type mismatch");
  if (data.sizeBytes <= 0) throw new AssetValidationError("invalid file size");
  const asset = await assetsRepo.confirm(projectId, data.assetId, {
    sizeBytes: data.sizeBytes,
    width: data.width,
    height: data.height,
  });
  return withPublicUrl(asset)!;
}

export async function listAssets(projectId: string): Promise<Record<string, unknown>[]> {
  await requireProject(projectId);
  const assets = await assetsRepo.listForProject(projectId);
  return assets.map((a) => withPublicUrl(a)!);
}

function randomId(): string {
  const buf = new Uint8Array(6);
  crypto.getRandomValues(buf);
  return Array.from(buf, (b) => b.toString(16).padStart(2, "0")).join("");
}
